cmds/tail/tail.go

```diff
@@ -0,0 +1,261 @@
+	// get the file's block size. This will be the size of each chunk to read
+	blkSize, err := getBlockSize(inFile)
+	if err != nil {
+		return err
+	}
+	// go to the end of the file
+	lastPos, err := inFile.Seek(0, os.SEEK_END)
+	if err != nil {
+		return err
+	}
+	// read block by block backwards until `numLines` lines are found
+	readData := make([]byte, 0)
+	buf := make([]byte, blkSize)
+	pos := lastPos
+	// for each block, count how many new lines, until they add up to `numLines`
+	for {
+		if pos == 0 {
+			break
+		}
+		var thisChunkSize int64
+		if pos < blkSize {
+			thisChunkSize = pos
+		} else {
+			thisChunkSize = blkSize
+		}
+		pos, err = inFile.Seek(-thisChunkSize, os.SEEK_CUR)
+		if err != nil {
+			return err
+		}
+		n, err := io.ReadFull(&inFile, buf)
+		if err != nil && err != io.ErrUnexpectedEOF {
+			return err
+		}
+		pos, err = inFile.Seek(-thisChunkSize, os.SEEK_CUR)
+		if err != nil {
+			return err
+		}
+		// merge this block to what read so far
+		readData = append(buf[:n], readData...)
+		// XXX this can be optimized, it's not necessary to re-count every time,
+		// just the last block
+		if bytes.Count(readData, []byte{'\n'}) >= numLines {
+			break
+		}
+	}
+	// find the start of the n-th to last line
+	slice := readData
+	// `data` contains up to `numLines` lines of the file
+	var data []byte
+	if len(slice) != 0 {
+		if slice[len(slice)-1] == '\n' {
+			// don't consider the last new line for the line count
+			slice = slice[:len(slice)-1]
+		}
+		var (
+			foundLines = 0
+			idx        int
+		)
+		for {
+			if foundLines >= numLines {
+				break
+			}
+			// find newlines backwards from the end of `slice`
+			idx = bytes.LastIndexByte(slice, '\n')
+			if idx == -1 {
+				// there are less than `numLines` lines
+				break
+			}
+			foundLines++
+			slice = slice[:idx-1]
+		}
+		if idx == -1 {
+			// if there are less than `numLines` lines, use all what we have read
+			data = readData
+		} else {
+			data = readData[idx+1:] // +1 to skip the newline belonging to the previous line
+		}
+	}
+	// write the requested lines to the writer
+	_, err = writer.Write(data)
+	if err != nil {
+		return err
+	}
+	// reposition the stream at the end of what we've read
+	_, err = inFile.Seek(lastPos, os.SEEK_SET)
+	return nil
+}
+
+// readLastLinesFromBeginning reads the last N lines from the provided file,
+// reading from the beginning of the file and keeping track of the last N lines.
+// This is necessary for files that are not seekable (e.g. stdin), but it's less
+// efficient. For an efficient alternative that works on seekable files see
+// readLastLinesBackwards.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesFromBeginning(inFile os.File, writer io.Writer, numLines int) error {
+	// get the file's block size. This will be the size of each chunk to read
+	blkSize, err := getBlockSize(inFile)
```

Comment:
block size seems like a premature optimization, but I see no harm in it either?

Reply:
that's what I thought being the most efficient approach in terms of syscalls, but I will instead use what suggested by @rminnich (i.e. blocks of size `81 * numLines`)